### PR TITLE
start dusd-api

### PR DESCRIPTION
**Pull request by Geometer1729**
Starts the dusd-api module with an updatable config utxo validator

- [ ] Every newly introduced function needs to have a [documentation](https://github.com/purescript/documentation/blob/master/language/Syntax.md#comments)
- [ ] The impure tests via `nix run .#offchain:test` were run and all are passing.
- [ ] Where applicable, grammar, punctuation, and spelling within code and comments should be plausibly correct.
- [ ] The issue(s) that the work is addressing should be linked in the PR description.
- [ ] Any significant changes beyond what's clear from the linked issue(s) are documented in the PR description

### CODE REVIEW

**Comment by Geometer1729** [comment]
@marijanp yeah, I guess we could make the ctl-utils module thing. The Dev experience is pretty annoying though...
Maybe just sim link the directory?

**Review by marijanp — APPROVED** [review]
LGTM, I want a squash merge on this, as most of the commits are not useful

File: offchain/dusd-api/src/Dusd/Api.purs [hunk 1]
```diff
@@ -0,0 +1,145 @@
+module Dusd.Api
```
[the remark this answers is not in the document]
Comment (Geometer1729) [reply offchain/dusd-api/src/Dusd/Api.purs:1]: :man_shrugging:

[the remark this answers is not in the document]
Comment (Geometer1729) [reply offchain/dusd-api/src/Dusd/Api.purs:1]: Dusd seems easiest to me, but this doesn't really matter. If you two can agree on something I'll go with it.

[the remark this answers is not in the document]
Comment (epiro-ps) [reply offchain/dusd-api/src/Dusd/Api.purs:1]: I'll yield to DUsd if you don't favor DUSD @marijanp

[the remark this answers is not in the document]
Comment (marijanp) [reply offchain/dusd-api/src/Dusd/Api.purs:1]: > It is an open question on how to name acronyms using this scheme. Some prefer editSGMLDocument while others prefer editSgmlDocument. The former is arguably more accurate and certainly more popular, but can get ugly
https://wiki.haskell.org/Camel_case

So `Dusd` is no valid camelcase

File: offchain/dusd-api/src/Dusd/Api.purs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+import Data.Array (cons)
+import Data.BigInt as BigInt
+import Data.List (head)
+import Data.Map (keys, singleton)
+import Data.Map as Map
+import Data.Set (toUnfoldable)
+import Dusd.CborTyped (configAddressValidator, simpleNft)
+import Effect.Exception (throw)
+
+newtype Protocol = Protocol { datum :: PlutusData, utxo :: TransactionInput, nftCs :: CurrencySymbol, configVal :: Validator }
```
Comment (marijanp) [inline offchain/dusd-api/src/Dusd/Api.purs:38]: Can we have the aeson instances here already?

**Review by marijanp — CHANGES_REQUESTED** [review]

File: offchain/dusd-api/src/Dusd/Api.purs (hunk 2 above)
Comment (Geometer1729) [reply offchain/dusd-api/src/Dusd/Api.purs:38]: Sure, I'll set that up.

**Review by marijanp — APPROVED** [review]
We'll stick with having ctl-utils inside the dusd-api. Right now we need to be pragmatic. Still want a squash merge

File: offchain/dusd-api/src/DUsd/CborTyped.purs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import Contract.Value (CurrencySymbol)
+import Effect.Exception (throw)
+
+{- This module should be the only place where CBOR is imported
+- all of its exports should handle all of the validator's parameters
+- this way there is only one module that needs to be checked
+- for type errors between on and off chain code
+-}
+
+configAddressValidator :: PubKeyHash -> CurrencySymbol -> Contract () Validator
```
Comment (epiro-ps) [inline offchain/dusd-api/src/DUsd/CborTyped.purs:25]: Could you add a doc here?

File: offchain/dusd-api/src/DUsd/Api.purs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        (Value.singleton nftCs adaToken one)
+    )
+  logDebug' "config utxo submitted, waiting for confirmation"
+  utxo <- waitForTx (scriptHashAddress $ validatorHash configVal) txid
+  logDebug' "protocol init complete"
+  pure $ Protocol { datum, utxo, nftCs, configVal }
+
+-- | technically not part of this version of the protocol
+updateProtocl :: PlutusData -> Protocol -> Contract () Protocol
+updateProtocl new (Protocol { utxo: oldUtxo, nftCs, configVal }) = do
```
Comment (epiro-ps) [inline offchain/dusd-api/src/DUsd/Api.purs:94]: *`updateProtocol`

**Review by epiro-ps — CHANGES_REQUESTED** [review]
Looks great. Just a few small things.

**Pull request merged**